src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -145,13 +146,13 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
 
                 if !ty_is_partial_eq {
                     // span_fatal avoids ICE from resolution of non-existent method (rare case).
-                    self.tcx().sess.span_fatal(self.span, &make_msg());
+                    self.tcx().sess.span_fatal(self.span, &msg);
```

Comment:
Doesn't this stop your ICE as well?

Reply:
nope. `ty_is_partial_eq` is `true` for `&dyn PartialEq<u32>`.  It is fixed for all other traits though.

This fails correctly
```rust
const F: &'static dyn Send = &7u32;

fn main() {
    let a: &dyn Send = &7u32;
    match a {
        F => panic!(),
        //~^ ERROR trait object cannot be used in patterns
        _ => {}
    }
}
```

Reply:
> It is fixed for all other traits though.

Ah I see. I think it's fine to leave `dyn PartialEq` an ICE (and we can do an ICE test, I think we have another one already). At least that's what this comment makes me think: https://github.com/rust-lang/rust/blob/32fb4dcdd7a57683a639a0959442711d0fd123bc/src/librustc_mir_build/hair/pattern/_match.rs#L317

@oli-obk has a PR that should allow fixing this more directly: #70743

Reply:
I changed the test to `dyn Send` and rebased.

Don't know how to test for an ICE, we could just open an issue for `dyn PartialEq` after this is merged (which is then added to glacier).